bin/elasticurl/main.cpp

```diff
@@ -0,0 +1,523 @@
+    semaphore.wait(semaphoreULock, [&]() { return connection || connectionShutdown; });
+
+    /* Send request */
+    int responseCode = 0;
+
+    Http::HttpRequest request;
+    Http::HttpRequestOptions requestOptions;
+    requestOptions.request = &request;
+
+    bool streamCompleted = false;
+    requestOptions.onStreamComplete = [&](Http::HttpStream &stream, int errorCode) {
+        (void)stream;
+        std::lock_guard<std::mutex> lockGuard(semaphoreLock);
+
+        streamCompleted = true;
+        if (errorCode)
+        {
+            errorOccured = true;
+        }
+
+        semaphore.notify_one();
+    };
+    requestOptions.onIncomingHeadersBlockDone = nullptr;
+    requestOptions.onIncomingHeaders = [&](Http::HttpStream &stream,
+                                           enum aws_http_header_block header_block,
+                                           const Http::HttpHeader *header,
+                                           std::size_t len) {
+        /* Ignore informational headers */
+        if (header_block == AWS_HTTP_HEADER_BLOCK_INFORMATIONAL)
+        {
+            return;
+        }
+
+        if (appCtx.IncludeHeaders)
+        {
+            if (!appCtx.ResponseCodeWritten)
+            {
+                responseCode = stream.GetResponseStatusCode();
+                std::cout << "Response Status: " << responseCode << std::endl;
+                appCtx.ResponseCodeWritten = true;
+            }
+
+            for (size_t i = 0; i < len; ++i)
+            {
+                std::cout.write((char *)header[i].name.ptr, header[i].name.len);
+                std::cout << ": ";
+                std::cout.write((char *)header[i].value.ptr, header[i].value.len);
+                std::cout << std::endl;
+            }
+        }
+    };
+    requestOptions.onIncomingBody = [&](Http::HttpStream &, const ByteCursor &data) {
+        if (appCtx.OutPut.is_open())
+        {
+            appCtx.OutPut.write((char *)data.ptr, data.len);
+        }
+        else
+        {
+            std::cout.write((char *)data.ptr, data.len);
+        }
+    };
+
+    request.SetMethod(ByteCursorFromCString(appCtx.verb));
+    request.SetPath(appCtx.uri.GetPathAndQuery());
+
+    Http::HttpHeader hostHeader;
+    hostHeader.name = ByteCursorFromCString("host");
+    hostHeader.value = appCtx.uri.GetHostName();
+    request.AddHeader(hostHeader);
+
+    Http::HttpHeader userAgentHeader;
+    userAgentHeader.name = ByteCursorFromCString("user-agent");
+    userAgentHeader.value = ByteCursorFromCString("elasticurl_cpp 1.0, Powered by the AWS Common Runtime.");
+    request.AddHeader(userAgentHeader);
+
+    std::shared_ptr<Io::StdIOStreamInputStream> bodyStream =
+        MakeShared<Io::StdIOStreamInputStream>(allocator, appCtx.InputBody, allocator);
+    int64_t dataLen = 0;
+    if (aws_input_stream_get_length(bodyStream->GetUnderlyingStream(), &dataLen))
```

Comment:
we shouldn't need to resort to the C API.
Let's either:
1) get the length from appCtx.InputBody
2) Add the public API to Aws::Crt::Io::InputStream: Seek(), Read(), GetStatus(), GetLength()

Reply:
did the second option